src/Build/Graph/ProjectGraph.cs

```diff
@@ -22,21 +28,69 @@ internal sealed class ProjectGraph
         /// <exception cref="InvalidProjectFileException">If the evaluation of any project in the graph fails.</exception>
         public ProjectGraph(string entryProjectFile)
         {
-            var graphNode =
-                new ProjectGraphNode(ProjectCollection.GlobalProjectCollection.LoadProject(entryProjectFile));
-            _projectNodes.Add(graphNode);
+            AddNewNode(entryProjectFile);
         }
 
         /// <summary>
         /// Constructs a graph starting from the given project files.
         /// </summary>
         /// <param name="entryProjectFiles">The project files to use as the entry points in constructing the graph</param>
         /// <exception cref="InvalidProjectFileException">If the evaluation of any project in the graph fails.</exception>
-        public ProjectGraph(IEnumerable<string> entryProjectFiles) { }
+        public ProjectGraph(IEnumerable<string> entryProjectFiles)
+        {
+            var projectsToParse = new Queue<string>();
+            foreach (var entryProjectFile in entryProjectFiles)
+            {
+                projectsToParse.Enqueue(entryProjectFile);
+            }
+
+            LoadGraph(projectsToParse);
+        }
 
         /// <summary>
         /// Get an unordered collection of all project nodes in the graph.
         /// </summary>
-        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _projectNodes;
+        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _allParsedProjects.Values;
+
+        private ProjectGraphNode AddNewNode(string projectFilePath)
+        {
+            var graphNode =
+                new ProjectGraphNode(ProjectCollection.GlobalProjectCollection.LoadProject(projectFilePath));
+            _allParsedProjects.Add(projectFilePath, graphNode);
+            return graphNode;
+        }
+
+        private List<ProjectGraphNode> LoadGraph(Queue<string> projectsToParse)
+        {
+            var parsedProjects = new List<ProjectGraphNode>();
```

Comment:
This seems to end up with the transitive closure, not just the direct project references.